entrypoint.py

```diff
@@ -83,6 +85,51 @@ async def pub_handler(request):
     return web.Response(text="success")
 
 
+# ws = new WebSocket("ws://${api_addr}/api/pub_ws")
+# ws.onopen = () => {
+#     ws.send(JSON.stringify({
+#         container: "...",
+#         topic: "...",
+#     }))
+# }
+# Then:
+# ws.send(JSON.stringify({
+#         container: "...",
+#         topic: "...",
+#         packet: {
+#             headers: [
+#                 ["key", "val"],
+#                 ...
+#             ],
+#             payload: window.btoa("..."),
+#         },
+#     }))
+async def pub_handler(request):
+    ws = web.WebSocketResponse()
+    await ws.prepare(request)
+
+    async for msg in ws:
+        if msg.type != WSMsgType.TEXT:
+            break
+
+        cmd = json.loads(msg.data)
+        global publishers
+
+        if cmd["topic"] in publishers:
+            publishers[cmd["topic"]].pub(a0.Packet(
+                cmd["packet"]["headers"],
+                base64.b64decode(cmd["packet"]["payload"])))
+        else:
+            tm = a0.TopicManager(container = cmd["container"])
+            publishers[cmd["topic"]] = a0.Publisher(tm.publisher_topic(cmd["topic"]))
+
+            publishers[cmd["topic"]].pub(a0.Packet(
+                cmd["packet"]["headers"],
+                base64.b64decode(cmd["packet"]["payload"])))
+
+        break
```

Comment:
Doesn't this unconditional break mean the websocket can only process one packet?

Reply:
I'm not sure. How would I be able to run and test this?

Reply:
the `async for msg in ws:` will break when the websocket is closed externally

Reply:
Cool. Fixed.